go/adbc/driver/databricks/record_reader.go

```diff
@@ -0,0 +1,312 @@
+
+		loadingChunkIdx: 0,
+		chunkChan:       make(chan chunkResponse),
+		activeChunk:     nil,
+
+		schema: nil, // TODO: build schema when there are no chunks
+		rec:    nil,
+		err:    nil,
+
+		cancelFn: func() {},
+
+		startTime: time.Now(),
+		// All the bytes received from the server.
+		BytesReceived: 0,
+		// Time spent waiting for the server to respond in the foreground.
+		WaitTime: 0,
+	}
+	if len(manifest.Chunks) > 0 && len(result.ExternalLinks) > 0 {
+		// Establish INVARIANT I by starting the loading of r.loadingChunkIdx
+		go r.startChunkDataRequest(r.loadingChunkIdx, &result.ExternalLinks[0])
+	} else {
+		// Establish INVARIANT I by finishing the entire iteration process
+		close(r.chunkChan)
+		r.loadingChunkIdx = -1
+	}
+	return r, nil
+}
+func (r *reader) Retain() {
+	atomic.AddInt64(&r.refCount, 1)
+}
+
+func (r *reader) Release() {
+	if atomic.AddInt64(&r.refCount, -1) == 0 {
+		if r.activeChunk != nil {
+			r.activeChunk.Release()
+		}
+		if r.rec != nil {
+			r.rec.Release()
+		}
+		// TODO: cancel HTTP connection
+		// TODO: close channel
+		r.cancelFn()
+	}
+}
+
+// \pre: loadingChunkIdx != -1 implies chunk is loading in the background (INVARIANT I)
+// \pre: loadingChunkIdx == -1 implies chunkChan is closed (INVARIANT II)
+// \post: if returns true, r.Record() != nil && r.err == nil
+// \post: if returns false, r.Record() == nil and r.err *MUST* be checked
+func (r *reader) Next() bool {
+	if r.rec != nil {
+		r.rec.Release()
+		r.rec = nil
+	}
+	if r.err != nil {
+		return false // post-condition holds: r.rec == nil && r.err != nil
+	}
+	// PROPERTY I: r.rec == nil && r.err == nil
+
+	// If we don't have an active chunk, we need to wait for the loading one,
+	// parse it, and trigger a request for the next chunk to preserve
+	// invariants I and II.
+	if r.activeChunk == nil {
+		if r.loadingChunkIdx == -1 {
+			return false // post-condition holds because of PROPERTY I
+		}
+
+		chunk, err := r.consumeLoadingChunk()
+		if err != nil {
+			r.err = err
+			close(r.chunkChan)
+			return false // post-condition holds because of PROPERTY I
+		}
+		r.activeChunk = chunk
+		r.activeChunk.Retain()
+
+		// make sure r.schema is set when the first chunk is parsed if not yet
+		if r.schema != nil {
+			r.schema = r.activeChunk.reader.Schema()
+		}
+	}
+	// PROPERTY II: r.activeChunk != nil
+
+	if r.activeChunk.reader.Next() {
+		r.rec = r.activeChunk.reader.Record()
+		r.rec.Retain()
+		return true // post-condition holds: r.rec != nil
+	}
+	// make sure the error (if it exists) is retained
+	r.err = r.activeChunk.reader.Err()
+	// release the fully consumed (or err'd) chunk
+	r.activeChunk.Release()
+	r.activeChunk = nil
+	// PROPERTY III: r.activeChunk == nil
+
+	// Recursively call Next() to start processing another chunk or stopping.
+	// Iteration will either terminate (if r.err != nil) or an attempt will be made
+	// to load the next chunk (because r.activeChunk == nil) guaranteeing progress.
+	return r.Next()
```

Comment:
is it optimizing tail recursion or better to restructure in imp style

Reply:
No, `Next()` either never gets called recursively or it's called recursively once and then it returns without recursing anymore. The fully imperative style would be putting the entire function in a `while true { .. }` and removing this `return` statement which would be more confusing I think.

Reply:
thanks for the explanation. It was not clear that it would recurse once at most based on the comment.